### ISSUE TITLE
Implemented conversion to a "mapping"

### ISSUE THREAD

**Issue by jorisvandenbossche** [pygeos/pygeos#82]
Shapely has a `mapping` function (https://shapely.readthedocs.io/en/latest/manual.html#shapely.geometry.mapping), which can convert a geometry into a dict (in geojson form):

```
In [59]: from shapely.geometry import mapping, LineString

In [60]: mapping(LineString([(1, 1), (3, 4)])) 
Out[60]: {'type': 'LineString', 'coordinates': ((1.0, 1.0), (3.0, 4.0))}
```

In geopandas, this is for example used in the `iterfeatures` / `__geo_interface__` and so also in `to_json` (a method that can convert to a geojson like string, which doesn't use fiona / GDAL like `to_file`, this is already faster as `to_file`, and can eg be used for in memory serialization to geojson).

From profiling `iterfeatures` in geopandas (for the countries dataset, so for polygons/multipolygons), the `mapping` of the geometries takes around 75% of the total time converting the dataframe to a dict. 

So this `mapping` could be a target for (vectorized) optimization. Since it involves creating python dicts and lists, it might be better suited to do this in cython, and also given that, I am not fully sure it will necessarily give a big speedup (experiments will need to tell).

This is related to https://github.com/pygeos/pygeos/issues/65 in the sense that the same dict needs to be created for both.

**Comment by jorisvandenbossche** [comment]
Also the conversion the other way around will be useful to have a fast geojson reader.

**Comment by caspervdw** [comment]
I think that putting this in C does not give any performance increase as you are creating many Python objects (rich of lists of floats). 

We could easily make a pure Python implementation and np.vectorize it. 

Or is there any support for geojson conversion in GEOS? Doing that and deserializing into a dict might be a lot faster.

**Comment by aprilmay** [comment]
A fast/direct way to generate a geojson representation would be great for using with many visualization libraries (Mapbox, Leaflet, etc).
No idea if that's useful, but gdal has a function: https://gdal.org/api/vector_c_api.html#_CPPv418OGR_G_ExportToJson12OGRGeometryH

**Comment by mullenkamp** [comment]
What's the current recommended way of going into and out of a geojson structured dict?
Is it using shapely as an intermediary since shapely provides a lot of this functionality?

**Comment by caspervdw** [comment]
Yes I would use shapely as an intermediary indeed.

**Comment by jorisvandenbossche** [comment]
Indeed, for now using shapely is probably the easiest way. 

> Or is there any support for geojson conversion in GEOS? Doing that and deserializing into a dict might be a lot faster.

It seems that recently there was some work adding a GeoJSON reader/writer to GEOS: https://github.com/libgeos/geos/pull/419 
However, that doesn't yet seem to include a C API, but so would be an interesting option to explore.

**Comment by caspervdw** [comment]
Closing; this is solved by #413
